### PR TITLE
Look in parent classes when parsing signal and method identifiers

### LINKED ISSUE #54172: GDScript 2.0: Accessing inherited signals and functions as members result in an "Identifier not found" error
### Godot version

v4.0.dev.custom_build [aec383a83]

### System information

Arch Linux 5.14.14

### Issue description

Accessing any signal/method as a member returns a parse error when running the script.
Example:

parent.gd:
```
extends Node
class_name Parent

signal parent_signal
```

child.gd:
```
extends Parent

func ready():
	# Errors on execution.
	parent_signal.connect(some_function.bind())

func some_function():
	print("Some behaviour.")
```

See the MRP for a more "realistic" and commented example.

Although not showcased, it should be noted that the old syntax works fine (`Object::connect` and `Object::emit_signal`, `Object::call` and similar basically).

I tried the best I could to fix it myself, but I'm way too inexperienced with the GDScript engine to fix this issue, nonetheless I collected some interesting information that could help whoever is capable of putting their hands in the compiler:

There is proper autocompletion (when it doesn't show its usual quirks anyway), so it means that the parser itself is working fine enough and according to my research (read: banging my head on my keyboard with GDB open for a week) all it does is just parsing a single class (but not local classes, which are interpreted as some sort of member) without accounting for inheritance, which is resolved by the compiler itself.

The real issue seems to stem from this line in `gdscript_compiler.cpp`:
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L291.

If we follow this variable a bit we can find what this codegen.class is:
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L1885
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L2410
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L2579
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L2562

It's the parser class! This is the only case in this method (aside from global classes and lambdas) which uses it, as can be seen here in other ones:
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L242-L258
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L260-L262

So why isn't the script being used? The issue apparently relies on the fact that there isn't some proper "conversion" between a parsed class and a proper script object in this regard. The function responsible for this is `GDScriptCompiler::_parse_class_level` which does properly include the parent class' members as can be seen there:
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L2209

But going even deeper in this mess reveals that, despite being accessed as such, they are never added to the actual members of the script!
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L2315-L2330

Notice how the only variable changed from this operation is `GDScript::_signals`, which is what is actually used for `Object::connect` and `Object::emit_signal`, as they both call `GDScript::has_script_signal` in their logic:
https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript.cpp#L1066-L1079

Notice also how this logic "works around" this situation by calling its `base`'s own `has_script_signal` method.

Now this is where everything starts to get kinda fuzzy and, honestly, out of my capabilities. At this point I already debugged *a lot* this complex piece of code which, at least to me, looks like it has built up some old behaviour around some quirks that are, AFAICT, pretty much as old as Godot as we know it:
old stuff

I have no idea how signals are handled internally _at all_ with this new syntax. Are they objects now? Were they one all along? I tried looking at where the compiled bytecode went, but it started getting way too out of hand for me. I tried adding them as if they were actual properties, and while it allowed them to be properly compiled (probably in the wrong way) it quickly errored out with an `UNITIALIZED` error of some sort, so there's clearly more going on.

And that's it, this is where I gave up. I hope I didn't give out useless or wrong information and that I might have helped @vnen or whoever knows better how this (at least to my eyes) very complex module works.

This certainly isn't a "brief description" lol.

### Steps to reproduce

1. Open up the MRP in the editor
2. Start the project
3. Notice it erroring in `cool_vending_machine.gd`
4. Follow its comments to continue in the reproduction (commenting the broken lines basically)
5. Notice that accessing inherited methods/signals as members gives out parser errors about identifiers not being found ([although they come from the compiler](https://github.com/godotengine/godot/blob/b2ab5cb504cef1ffe229523d63b997d5306afce3/modules/gdscript/gdscript_compiler.cpp#L372))

### Minimal reproduction project

[signal-and-function-inheritance-issue-mrp.zip](https://github.com/godotengine/godot/files/7403864/signal-and-function-inheritance-issue-mrp.zip)

### LINKED ISSUE #57337: GDScript 2.0: "Identifier not found" when trying to connect to super function
### Godot version

v4.0.alpha1.official [31a7ddbf8]

### System information

windows 10

### Issue description

When when trying to connect a signal to super function, willl report "Identifier not found" when run.
ttt

Note: If add `self` before `f1`, it will run well.

### Steps to reproduce

1. Open the reproduction project.
2. Run

### Minimal reproduction project

[SuperFuncConnectTest.zip](https://github.com/godotengine/godot/files/7956464/SuperFuncConnectTest.zip)

### PR DESCRIPTION
**Pull request by Piet-G**
Fixes #57337

Added two tests for this scenario

*Bugsquad edit:*
- Fixes #54172

### CODE REVIEW

**Comment by Piet-G** [comment]
It seems this didn't pass the tests, but the error message is extremely unclear. I don't really see how my changes could have caused this. Does anyone have an idea?

File: modules/gdscript/tests/scripts/analyzer/features/parent_class_methods_indentified.gd [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+    func f2():
+        pass
+
+    func test_connect():
+        test_signal.connect(f1)
+        test_signal.connect(f2)
+
+func test():
+    B.new().test_connect()
```
Comment (akien-mga) [inline modules/gdscript/tests/scripts/analyzer/features/parent_class_methods_indentified.gd:18]: You're leaking `B` here, it needs to be freed (either manually, or by adding it to the scene tree).

File: modules/gdscript/tests/scripts/analyzer/features/parent_class_signals_indentified.gd [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+
+    func f1():
+        pass
+
+    func test_connect():
+        test_signal_a.connect(f1)
+        test_signal_b.connect(f1)
+
+func test():
+    B.new().test_connect()
```
Comment (akien-mga) [inline modules/gdscript/tests/scripts/analyzer/features/parent_class_signals_indentified.gd:17]: Also leaking `B`.

**Comment by akien-mga** [comment]
> It seems this didn't pass the tests, but the error message is extremely unclear. I don't really see how my changes could have caused this. Does anyone have an idea?

The leak sanitizer found that your tests are leaking memory (you instantiate a Node, which is not RefCounted, so it needs to be freed somehow).

**Comment by Piet-G** [comment]
> > It seems this didn't pass the tests, but the error message is extremely unclear. I don't really see how my changes could have caused this. Does anyone have an idea?
> 
> The leak sanitizer found that your tests are leaking memory (you instantiate a Node, which is not RefCounted, so it needs to be freed somehow).

Thanks for the help, should be fixed now.

**Comment by strank** [comment]
Hi there. There is another PR for the same problem: #56830 (and a duplicate corresponding issue #54172). I am the author of the other PR, so somewhat biased - they seem roughly equivalent so it would probably be good for whoever reviews to compare. I don't have unit tests in the other one, only a test file in the PR description. The main differences seem to be:

* the method of walking up the hierarchy: `scr->_base` vs. `base_class->base_type.class_type`  - I couldn't check easily if they are always equivalent.
* one collapses the `try in native base` loop with the new loop walking up the hierarchy. I am not completely sure why that is kept separate in the original code, so I left it separate.
* Use of `base_class->has_member` and `base_class->get_member` which I think improves the code readability.

Feel free to integrate as you see fit!

**Comment by vnen** [comment]
I actually prefer @strank's solution TBH. But I like that this PR includes a test case which helps to avoid future regressions.

**Comment by akien-mga** [comment]
We merged #56830 which solves the issue, but it would be good to add the test cases from this PR indeed. @Piet-G Could you rebase to keep the tests only?

**Comment by akien-mga** [comment]
Superseded by #56830 and #58757. Thanks dir the contribution nevertheless!

**Pull request closed without merging**